keps/sig-instrumentation/0034-distributed-tracing-kep.md

```diff
@@ -0,0 +1,179 @@
+
+
+## Motivation
+
+Debugging latency issues in Kubernetes is an involved process. There are existing tools which can be used to isolate these issues in Kubernetes, but these methods fall short for various reasons. For instance:
+
+* **Logs**: are fragmented, and finding out which process was the bottleneck involves digging through troves of unstructured text. In addition, logs do not offer higher-level insight into overall system behavior without an extensive background on the process of interest. 
+* **Events**: in Kubernetes are only kept for an hour by default, and don't integrate with visualization of analysis tools. To gain trace-like insights would require a large investment in custom tooling.
+* **Latency metrics**: are gathered in some places, but these don't provide understanding into _why_ a given process was slow.
+
+Distributed tracing, on the other hand, provides a single window into latency information from across many components and plugins. Trace data is structured, and there are numerous established backends for visualizing and querying over it.
+
+In addition, due to the self-healing nature of Kubernetes, regressions wherein latencies are affected but the overall task is eventually accomplished are not uncommon. With our current monitoring architecture, these "soft regressions" are often difficult to observe and understand. Collecting structured trace data on per-object latencies would enable us to detect these long-term regressions automatically, and quickly determine their root causes.
+
+
+### Goals
+
+* Make it possible to visualize the progress of objects across distinct Kubernetes components
+* Streamline the tedious processes involved in debugging latency issues in Kubernetes
+* Make it possible to identify high-level latency regressions, and attribute them back to offending processes
+
+
+### Non-Goals
+
+* To replace existing logging, metrics, or the events API
+* To trace operations from all Kubernetes resource types in a generic manner (i.e. without manual instrumentation)
+
+
+## Proposal
+
+### Architecture
+
+#### Trace lifecycle
+
+Kubernetes is unique in that it is constantly reconciling its actual state towards some desired state. As a result, it has no definitive concept of an "operation", which breaks the traditional model for distributed tracing. This raises the question of when to begin traces, and when to end them.
+
+In this proposal, we choose to _only_ trace phases of an object's lifecycle wherein it's correcting from an undesired state to its desired state, and to end the trace when it enters this desired state. This means that the same object will export traces for each reconciliation it undergoes. This decision was made because:
```

Comment:
What if in the meantime desired state changes and we will never reach the original desired one?

Reply:
The original trace will end prematurely, and subsequent traced actions are attributed to the new desired state.  Since we generally care about the slowest reconciliations, ending a trace before a reconciliation is complete is fine.

Reply:
Assuming that it's the component that has the knowledge about that previous trace...
Anyway - I think any option here is potentially fine, but I would like to see that written down in the KEP to give people chance to discuss that.